### ISSUE TITLE
Simple diffusion with QUAD8 and one uniform refinement fails after updating libmesh to 4775923

### ISSUE THREAD

**Issue by jwpeterson** [idaholab/moose#4966]
To replicate:  add
```
elem_type = QUAD8
uniform_refine = 1
```
to the [Mesh] block, and
```
order = SECOND
family = LAGRANGE
```
to the u-variable section of `test/tests/kernels/simple_diffusion/simple_diffusion.i`.  What's really weird to me about this error is that we call a function on a `RemoteElem` on a serial program running in serial!  @roystgnr it looks like we are down in `System::init_data()`, would this have anything to do with the refactored system projection code?  The same test works fine either *without* uniform refinement or using `QUAD9` elements.

```
Stack frames: 19
0: 0   libmesh_opt.0.dylib                 0x0000000108028965 libMesh::print_trace(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) + 1237
1: 1   libmesh_opt.0.dylib                 0x0000000108026f39 libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*) + 329
2: 2   libmesh_opt.0.dylib                 0x00000001081ea7d6 libMesh::RemoteElem::n_nodes() const + 38
3: 3   libmesh_opt.0.dylib                 0x0000000107fdc70b libMesh::DofMap::dof_indices(libMesh::Elem const*, std::__1::vector<unsigned int, std::__1::allocator<unsigned int> >&) const + 651
4: 4   libmesh_opt.0.dylib                 0x0000000107fdab98 libMesh::DofMap::add_neighbors_to_send_list(libMesh::MeshBase&) + 2424
5: 5   libmesh_opt.0.dylib                 0x0000000107fd903b libMesh::DofMap::distribute_dofs(libMesh::MeshBase&) + 1787
6: 6   libmesh_opt.0.dylib                 0x0000000108476655 libMesh::System::init_data() + 149
7: 7   libmesh_opt.0.dylib                 0x0000000108469be5 libMesh::ImplicitSystem::init_data() + 21
8: 8   libmesh_opt.0.dylib                 0x00000001084aa9f3 libMesh::TransientSystem<libMesh::NonlinearImplicitSystem>::init_data() + 19
9: 9   libmesh_opt.0.dylib                 0x0000000108476588 libMesh::System::init() + 40
10: 10  libmesh_opt.0.dylib                 0x000000010844b552 libMesh::EquationSystems::init() + 786
11: 11  libmoose-opt.0.dylib                0x00000001077f2e02 FEProblem::init() + 866
12: 12  libmoose-opt.0.dylib                0x0000000107703789 ActionWarehouse::executeActionsWithAction(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) + 313
13: 13  libmoose-opt.0.dylib                0x0000000107703602 ActionWarehouse::executeAllActions() + 274
14: 14  libmoose-opt.0.dylib                0x000000010789c062 MooseApp::runInputFile() + 210
15: 15  libmoose-opt.0.dylib                0x000000010789d168 MooseApp::run() + 24
16: 16  moose_test-opt                      0x0000000107432571 main + 113
17: 17  libdyld.dylib                       0x00007fff975805c9 start + 1
18: 18  ???                                 0x0000000000000003 0x0 + 3
```

**Comment by permcody** [comment]
After this is fixed, we should definitely add this to the test suite.

**Comment by jwpeterson** [comment]
I found an old checkout laying around, and confirm this example works with the given modifications as of libmesh/libmesh@c76d905, still need to do a more formal git bisection...

**Comment by roystgnr** [comment]
Thanks for the heads up; I'll see if I can replicate, both with MOOSE and with as similar as I can get (QUAD8, SECOND) in other codes.  I use the libMesh head heavily but I never use QUAD8 myself so that's almost certainly the issue.

And thinking about the problem, I think the prime suspect has got to be https://github.com/libMesh/libmesh/pull/492 - if you're bisecting then I'd try https://github.com/libMesh/libmesh/commit/0f9f71e069ddcd7d50a851c3a76f74a3f2701fba and https://github.com/libMesh/libmesh/commit/c76d905ae7949dc053d7baed562649da2b4501d3

**Comment by jwpeterson** [comment]
So you are thinking it is topology map stuff, and not the `GenericProjector` stuff?

This is consistent with what I just found, at least: the same error occurs in libmesh/libmesh@d22ceec, which is the commit immediately before the `GenericProjector` stuff was merged.

**Comment by roystgnr** [comment]
Definitely an easy to reproduce bug (just refine some QUAD8...), and definitely not a MOOSE bug; I get the same error with both MOOSE and GRINS.

**Comment by roystgnr** [comment]
RemoteElem (on SerialMesh!  in serial!?!) suggests to me that the find_neighbors loop isn't finding neighbors that should be there, which in turn suggests that we're inadvertently "dissecting" the mesh upon refinement, which would be a failure mode of the topology map if there's some case that my "generic" mapping code didn't account for.

**Comment by jwpeterson** [comment]
libmesh/libmesh@8c4dfbe, the commit immediately after the topology map merge, fails.

**Comment by roystgnr** [comment]
Yup...  QUAD8 (and HEX20...) is such a case.  I'll see if I can figure out a correction.

**Comment by jwpeterson** [comment]
And just to confirm, libmesh/libmesh@c5213a3 (the last commit on the `metis_fixes` branch before it was merged to master) works. So, the bug must have been introduced by the `use_topology_map` branch... would it be helpful for me to bisect on the `use_topology_map` branch itself or are you already doing that?

**Comment by roystgnr** [comment]
I'm already at work.  Bisection won't help, I'm afraid; you'll just find out that the point where LocationMap use got replace by TopologyMap use was the point that broke.  But I'm almost to the right spot (bracketing node cache generation) in the debugger and once there I'm sure I'll be able to figure out why it's missing the cache pairs I suspect it's missing.

**Comment by jwpeterson** [comment]
> Bisection won't help, I'm afraid; you'll just find out that the point where LocationMap use got replace by TopologyMap

You are right, FWIW I confirmed that libmesh/libmesh@8cbea78 is the first broken commit on the `use_topology_map` branch.

**Comment by roystgnr** [comment]
Okay, I've figured out what the precise problem is.  Let me write it out, if only for future reference and my own edification; I think I can handle the fix without help.

New nodes which are shared between child elements (e.g. in the QUAD8 case: node 5 on child 0 is node 7 on child 1) need to be uniquely defined.  LocationMap defined them via x,y,z, but this breaks on slit meshes where you don't *want* every pair of nodes at the same point to be unique.  TopologyMap defines them via "bracketing nodes": collections of pairs of global node ids on the parent element such that, for each such pair a,b, the new node is topologically right in between them (e.g. a midface node on a HEX27 child is bracketed by both pairs of opposite-corner nodes on that face) and is on the same submanifold as them (e.g. on a QUAD9, child 1 node 6 is *not* bracketed by parent nodes 1 and 6).

But this breaks in the QUAD8/HEX20 case, because the mid-edge nodes on new interior-to-parent-element edges are *not* bracketed by parent node pairs, but rather by pairs of parent-node-on-the-edge and missing-node-in-the-parent-center.

I *think* I can actually handle that, though...  I can have Elem::parent_bracketing_nodes() cache "255" or some other nonsense as a local node id for the case where no correct local node id exists.  Then when we run into "255" while translating from cached local node ids to element-specific global node ids in Elem::bracketing_nodes(), we can assume (and libmesh_assert() for good measure) that we've already created the node in question via child 0 (due to our vertex-first ordering and the fact that we assign temporary unique node ids immediately as each is created), and we can query via that node pointer to get the new center node id to add to the map.

The only catch is that we'll need to add Quad8::bracketing_nodes() and Hex20::bracketing_nodes() (and any other "weird" elements?  InfHex16... Prism15... that should do it) overrides, but that method is already virtual from back when I had the half-baked idea of hard-coding bracketing caches rather than computing them on first invocation.

**Comment by roystgnr** [comment]
Ah, InfQuad6 too.

**Comment by roystgnr** [comment]
Oh, hell.  My plan isn't going to be so simple for Hex20 or anything else with a Quad8 side, because in that case the missing-node-in-the-parent-center might instead be any one of several missing-nodes-in-the-parent-sides, and there needs to be some way for the overridden bracketing_nodes() to figure out which.

**Comment by jwpeterson** [comment]
Fixed in b54d009
